src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -230,48 +248,49 @@ public class EpochManager : IEpochManager {
 
 		public void SetLastEpoch(EpochRecord epoch) {
 			Ensure.NotNull(epoch, "epoch");
+			//if in follower/clone mode we should make sure the replicated epoch is in the manager cache
+			//this method is idempotent and safe to be called on the leader as well
 
-			lock (_locker) {
-				if (epoch.EpochPosition > _lastEpochPosition) {
-					UpdateLastEpoch(epoch, flushWriter: false);
-					return;
-				}
+			if (epoch.EpochPosition > LastEpochNumber) {
+				UpdateLastEpoch(epoch, flushWriter: false);
+				return;
 			}
 
 			// Epoch record must have been already written, so we need to make sure it is where we expect it to be.
 			// If this check fails, then there is something very wrong with epochs, data corruption is possible.
 			if (!IsCorrectEpochAt(epoch.EpochPosition, epoch.EpochNumber, epoch.EpochId)) {
-				throw new Exception(string.Format("Not found epoch at {0} with epoch number: {1} and epoch ID: {2}. "
-				                                  + "SetLastEpoch FAILED! Data corruption risk!",
-					epoch.EpochPosition,
-					epoch.EpochNumber,
-					epoch.EpochId));
+				throw new Exception(
+					$"Not found epoch at {epoch.EpochPosition} with epoch number: {epoch.EpochNumber} and epoch ID: {epoch.EpochId}. " +
+					"SetLastEpoch FAILED! Data corruption risk!");
 			}
 		}
 
 		private void UpdateLastEpoch(EpochRecord epoch, bool flushWriter) {
+			if (LastEpochNumber == epoch.EpochNumber) {
+				return;
+			}
+			if (LastEpochNumber > epoch.EpochNumber) {
+				throw new Exception($"Concurrency failure, Last Epoch Number is {LastEpochNumber} cannot add new Last Epoch with a lower Epoch Number {epoch.EpochNumber}.");
+			}
+
 			lock (_locker) {
-				_epochs[epoch.EpochNumber] = epoch;
-				_lastEpochNumber = epoch.EpochNumber;
-				_lastEpochPosition = epoch.EpochPosition;
-				_minCachedEpochNumber = Math.Max(_minCachedEpochNumber, epoch.EpochNumber - CachedEpochCount + 1);
-				_epochs.Remove(_minCachedEpochNumber - 1);
-
-				if (flushWriter)
-					_writer.Flush();
-				// Now update epoch checkpoint, so on restart we don't scan sequentially TF.
-				_checkpoint.Write(epoch.EpochPosition);
-				_checkpoint.Flush();
-
-				Log.Debug(
-					"=== Update Last Epoch E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}) L={leaderId:B}.",
-					epoch.EpochNumber, epoch.EpochPosition, epoch.EpochId, epoch.PrevEpochPosition, epoch.LeaderInstanceId);
+				if (!_epochs.Contains(epoch)) {
+					_epochs.AddLast(epoch);
+					_lastCachedEpoch = _epochs.Last;
+					_epochs.RemoveFirst();
```

Comment:
@condron We should remove an item only if the linked list size was exactly equal to _cacheSize